src/vm.rs

```diff
@@ -136,7 +71,59 @@ pub trait VirtualCPU {
 
 	fn args(&self) -> &[OsString];
 
-	fn cmdsize(&self, syssize: &mut SysCmdsize) {
+	/// `addr` is the address of the hypercall parameter in the guest's memory space. `data` is the
+	/// parameter that was send to that address by the guest.
+	fn address_to_hypercall(&self, addr: u16, data: usize) -> Option<Hypercall<'_>> {
```

Comment:
Ideally, this would also take `data: *mut ()` at least for provenance, but changing that might be out of scope for this PR, depending on how hard that would collide with the rest of the code.

Reply:
I see your point, but that is what the interface was in the first place. So, I would suggest reworking this in a later PR.

Reply:
How has this been the interface? As far as I can see, you introduced this function in this PR.

Previously, the two points were fine.

The casting happened in a `&mut self` context before, and the lifetime of the structs were restricted by the scope. I think, these two points should really be addressed.

Another thing: This has to be `unsafe`, since we require `addr` and `data` to be valid and create references from pointers derived from these values.

Reply:
Ok, got it.
I made the fn `unsafe` and added safety doc.
I'd suggest creating a "making it safer" issue and do it separately.  I'm experimenting with a newtype for guest addresses anyway, and this could have some synergies.